### PR TITLE
Retry implementation for LoadBalanced RestTemplates

### PR DESCRIPTION
**Pull request by ryanjbaxter**
Helps address spring-cloud/spring-cloud-netflix#1290

### CODE REVIEW

**Comment by william-tran** [comment]
How would I specify retry on 404 behaviour? Is there still the issue of the underlying client not throwing on such errors and so not triggering retry behaviour?

**Comment by ryanjbaxter** [comment]
Right now there is no retry logic with load balanced `RestTemplates` so the existing behavior never retries.  Unlike with the Netflix `RestClient` this will retry on any type of error.

File: spring-cloud-commons/pom.xml [hunk 1]
```diff
@@ -54,6 +54,14 @@
 			<optional>true</optional>
 		</dependency>
 		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-aop</artifactId>
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.retry</groupId>
+			<artifactId>spring-retry</artifactId>
```
Comment (spencergibb) [inline spring-cloud-commons/pom.xml]: Should these be optional?

Comment (ryanjbaxter) [inline spring-cloud-commons/pom.xml]: All set

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancerAutoConfiguration.java [hunk 2]
```diff
@@ -25,9 +25,11 @@
 import org.springframework.boot.autoconfigure.condition.ConditionalOnBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
 import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.retry.support.RetryTemplate;
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancerAutoConfiguration.java]: I guess when I asked if the dependencies should be optional, that implied that this should work without retry. Either it should work without retry or the depencencies shouldn't be optional. @dsyer?

Comment (dsyer) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancerAutoConfiguration.java]: I like optional.

Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancerAutoConfiguration.java]: Maybe it is better to just move the load balancer code to another module?

Comment (dsyer) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancerAutoConfiguration.java]: Why?

Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancerAutoConfiguration.java]: Not sure I understand how you would remove the imports for spring retry and still add the functionality to retry the request.  

If we moved the load balancer code to its own module in commons than that module can depend on spring retry.  If downstream user's of commons dont need the load balancer stuff than they would not get the spring retry dependency.

Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalancerAutoConfiguration.java]: @spencergibb @dsyer with the latest changes spring-retry should be truly optional

File: spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalanceChooser.java [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+import org.springframework.cloud.client.ServiceInstance;
+
+/**
+ * Implemented by classes which use a load balancer to choose a server to
+ * send a request to.
+ *
+ * @author Ryan Baxter
+ */
+public interface LoadBalanceChooser {
```
Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalanceChooser.java]: Maybe a better name? `ServiceInstanceChooser`?

Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalanceChooser.java]: Actually, why the separation? Something not implemented yet? I don't see it in s-c-netflix.

Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalanceChooser.java]: `LoadBalancerClient` extends this interface.  `RibbonLoadBalancerClient` in s-c-netflix implements `LoadBalancerClient`.

Comment (ryanjbaxter) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalanceChooser.java]: I am fine with renaming it to `ServiceInstanceChooser`.

Comment (spencergibb) [inline spring-cloud-commons/src/main/java/org/springframework/cloud/client/loadbalancer/LoadBalanceChooser.java]: cool, I'd say that's the last thing.

**Comment by ryanjbaxter** [comment]
@spencergibb since spring-cloud/spring-cloud-netflix#1375 is related I will wait for your approval there as well before merging this.  Thanks!

**Pull request merged**